data_cache.py

```diff
@@ -0,0 +1,117 @@
+        :return: tuple of (dict, bool) where dict is cached location (or None) and bool is whether location is valid
+        """
+        # setup index if it doesnt already exist
+        self.collection.create_index([('location', GEO2D)])
+
+        # find nearest location
+        nearest_cached_loc = self.collection.find_one({'location': {'$near': [lng, lat]}})
+
+        # check if valid cache object is returned
+        if nearest_cached_loc is not None:
+            # compute distance to nearest cached object
+            nearest_cached_loc_location = (nearest_cached_loc['location'][1], nearest_cached_loc['location'][0])
+            dist_to_nearest = vincenty(nearest_cached_loc_location, (lat, lng)).meters
+
+            # compute time diff between cached object and current time
+            current_date = datetime.datetime.utcnow()
+            time_delta_sec_to_nearest = (current_date - nearest_cached_loc['date']).total_seconds()
+            time_delta_mins_to_nearest = divmod(time_delta_sec_to_nearest, 60)[0]
+
+            print('{} -- Nearest cached location: {} meters away, {} minutes ago.'.format(self.collection_name,
+                                                                                          dist_to_nearest,
+                                                                                          time_delta_mins_to_nearest))
+
+            # cache is valid if within distance
+            if dist_to_nearest < self.distance_threshold:
+                # return cache object iff valid AND within time threshold
+                if time_delta_mins_to_nearest < self.time_threshold:
+                    return nearest_cached_loc, True
+                else:
+                    return nearest_cached_loc, False
+
+        # no valid cache object could be found
+        return None, False
+
+    def add_to_cache(self, lat, lng, data_to_save):
+        """
+        Adds location to cache.
+
+        :param lat: Latitude of location, as float.
+        :param lng: Longitude of location, as float.
+        :param data_to_save: Data to save for location, as list
+        :return: inserted id of document, if successful
+        """
+        return self.collection.insert_one({
+            'location': [lng, lat],  # longitude, latitude format
+            'data': data_to_save,
+            'date': datetime.datetime.utcnow()
+        }).inserted_id
+
+    def update_cache(self, object_id, new_data_to_save):
+        """
+        Updates existing location in cache.
+
+        :param object_id: Id of object to update, as ObjectId
+        :param new_data_to_save: Data to save for location, as list
+        :return: inserted id of document, if successful
+        """
+        return self.collection.update_one({
+            '_id': object_id
+        }, {
+            '$set': {
+                'data': new_data_to_save,
+                'date': datetime.datetime.utcnow()
+            }
+        }, upsert=True)
```

Comment:
Do we really want `upsert=True` here?  If upsert is invoked, the document in the collection will not have `location: [lng, lat]` values.

Reply:
nope, good catch. This will also only be run when there's a valid object so no need to upsert ever